reps/2023-08-18-serve-java-dag-api.md

```diff
@@ -0,0 +1,266 @@
+## Summary
+### General Motivation
+Compared to the latest Serve API in Python, Ray Serve Java lacks a major version update. Currently, the main form of the Serve API is the DAG API, which involves networking deployments through binding and deploying the graph. For example:
+```python
+import ray
+from ray import serve
+from ray.serve.dag import InputNode
+from ray.serve.drivers import DAGDriver
+
+
+@serve.deployment
+def preprocess(inp: int) -> int:
+    return inp + 1
+
+
+@serve.deployment
+class Model:
+    def __init__(self, increment: int):
+        self.increment = increment
+
+    def predict(self, inp: int) -> int:
+        return inp + self.increment
+
+
+with InputNode() as inp:
+    model = Model.bind(increment=2)
+    output = model.predict.bind(preprocess.bind(inp))
+    serve_dag = DAGDriver.bind(output)
+
+handle = serve.run(serve_dag)
+assert ray.get(handle.predict.remote(1)) == 4
+
+```
+Looking back at Java, we hope that Java can keep up with the features of the Serve API, allowing Java developers to deploy Java projects as Serve deployments and compose multiple deployments to accomplish complex online computations.
+### Should this change be within `ray` or outside?
+Main `ray` project. A part of java/serve.
+## Stewardship
+
+### Required Reviewers
+@sihanwang41
+
+### Shepherd of the Proposal (should be a senior committer)
+@sihanwang41
+
+## Design and Architecture
+
+### Update Java user API to be consistent with Python
+A standard Java deployment demo is shown below:
+```java
+// Demo 1
+import io.ray.serve.api.Serve;
+import io.ray.serve.deployment.Application;
+import io.ray.serve.handle.RayServeHandle;
+
+public class DeploymentDemo {
+  private String msg;
+
+  public DeploymentDemo(String msg) {
+    this.msg = msg;
+  }
+
+  public String call() {
+    return msg;
+  }
+
+  public static void main(String[] args) {
+    Application deployment =
+        Serve.deployment().setDeploymentDef(DeploymentDemo.class.getName()).bind();
+    RayServeHandle handle = Serve.run(deployment).get();
+    System.out.println(handle.remote().get());
+  }
+}
+
+```
+In this demo, a DAG node is defined through the `bind` method of the Deployment, and it is deployed using the `Serve.run` API.
+Furthermore, a Deployment can bind other Deployments, and users can use the Deployment input parameters in a similar way to `RayServeHandle`. For example:
+```java
+// Demo 2
+import io.ray.api.ObjectRef;
+import io.ray.serve.api.Serve;
+import io.ray.serve.deployment.Application;
+import io.ray.serve.handle.RayServeHandle;
+
+public class Driver {
+  private RayServeHandle modelAHandle;
+  private RayServeHandle modelBHandle;
+
+  public Driver(RayServeHandle modelAHandle, RayServeHandle modelBHandle) {
+    this.modelAHandle = modelAHandle;
+    this.modelBHandle = modelBHandle;
+  }
+
+  public String call(String request) {
+    ObjectRef<Object> refA = modelAHandle.remote(request);
+    ObjectRef<Object> refB = modelBHandle.remote(request);
```

Comment:
in the most recent Python API we are not returning an ObjectRef directly but have a wrapping object called a `DeploymentResponse`. we might consider doing the same for java.

the main reason for doing it in python is to avoid having to "double await" to get the final result. maybe that's less relevant here.

Reply:
Alright, I have made the adjustments according to the suggestions.